EDAC: Fixing API and API docs generation, fixing static analyzer warnings

finikorg opened the pull request:
Make it possible to generate full EDAC API for documentation.
Fixing warnings from static analyzer.

finikorg committed:
edac: Introduce enum type for error type

Use edac_error_type enum type instead of defines for EDAC error types.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: Refactor EDAC API

Refactor EDAC API making it more clear, removing unneeded typedefs and
using check for optional and assert for mandatory APIs.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: shell: Use new EDAC API in shell

Use new EDAC API in shell.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: tests: Use new EDAC API in tests

Use new EDAC API in tests.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: Include error injection API to documentation

Remove #ifdefs making code cleaner and including error injection API
functions to documentation.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: docs: Use @retval instead of @return when applicable

Make documentation more readable by using @retval instead of @return
when applicable.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

nashif commented on drivers/edac/edac_ibecc.c:
```diff
@@ -207,9 +207,11 @@ static int inject_set_param1(const struct device *dev, uint64_t addr)
 }
 
-static uint64_t inject_get_param1(const struct device *dev)
+static int inject_get_param1(const struct device *dev, uint64_t *value)
```
this violates misra dir 4.6: "typedefs that indicate size and signedness should be used in place of the basic numerical types"

Cane we please keep using typedefs instead.

finikorg replied:
@nashif , sorry do not get this. with respect to this change above typedef does not matter. I return here `int` instead of value to be able to return `-ENOTSUP`.

jhedberg replied:
If the intent is to return `0` on success and `-errno` on failure then `int` is the appropriate choice, I believe.

nashif replied:
> If the intent is to return `0` on success and `-errno` on failure then `int` is the appropriate choice, I believe.

directive 4.6 says otherwise though.

jhedberg replied:
I guess we might need to state an explicit exception to this rule in our coding style conventions then, since I believe `int` return is used in lots of places of Zephyr, in particular in the core kernel code. If I instead see e.g. `int32_t` I will think that the return value has some other meaning than the usual "0 or -errno", whereas if I see a `int` return value I can assume that it follows the posix error number semantics. I think this is a valuable convention to keep around since that way we keep it separate from some use cases that really do require a fixed-width integer.

finikorg replied:
@nashif I did not understood first that you are talking about this change `int` -> `int32_t`. I thought the comment was about removing some typedefs above. I can change `int` to `int32_t`, although, as @jhedberg mentioned, this is (was?) common for return type in Zephyr API.

finikorg replied:
@nashif Added 617c528  converting `int` to `int32_t`.

sjg20 replied:
IMO this line is correct. We should use int (i.e. the natural size of the machine) for return codes, not int32_t or int64_t

nashif commented on include/drivers/edac.h:
```diff
@@ -239,29 +266,38 @@ static inline void edac_ecc_error_log_clear(const struct device *dev)
 		(const struct edac_driver_api *)dev->api;
 
-	return api->parity_error_log_get(dev);
+	__ASSERT(api->parity_error_log_get,
```
this should return -ENOSYS, see https://docs.zephyrproject.org/latest/guides/design_guidelines.html#return-codes

finikorg replied:
Updated in 65f0dc2

finikorg committed:
edac: Use -ENOSYS for not implemented mandatory API functions

Use standard Zephyr way for not implemented mandatory API functions.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: Use TODO instead of workaround in comment

Using TODO is more appropriate since workaround implies working around
a bug.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jhedberg commented on drivers/edac/shell.c:
```diff
@@ -62,6 +62,7 @@ static int cmd_edac_info(const struct shell *shell, size_t argc, char **argv)
 {
 	const struct device *dev;
 	uint64_t error;
+	int ret;
```
Please use `err` for simple `int` type errors. If you think this will be confusing with your `error` variable, then rename that one e.g. to `ecc_err` or something similar.

finikorg replied:
done

jhedberg commented on drivers/edac/shell.c:
```diff
@@ -71,27 +72,38 @@ static int cmd_edac_info(const struct shell *shell, size_t argc, char **argv)
 static int cmd_inject_addr(const struct shell *shell, size_t argc, char **argv)
 {
 	const struct device *dev;
+	int ret;
```
Same here

jhedberg commented on drivers/edac/shell.c:
```diff
@@ -116,15 +134,21 @@ static int cmd_inject_addr(const struct shell *shell, size_t argc, char **argv)
 static int cmd_inject_mask(const struct shell *shell, size_t argc, char **argv)
 {
 	const struct device *dev;
+	int ret;
```
And here

jhedberg commented on drivers/edac/shell.c:
```diff
@@ -71,27 +72,38 @@ static int cmd_edac_info(const struct shell *shell, size_t argc, char **argv)
 
-	error = edac_ecc_error_log_get(dev);
+	ret = edac_ecc_error_log_get(dev, &error);
+	if (ret) {
```
Assuming this is under the scope that the coding style will be enforced on, I think only boolean expressions or bool type variables can be used as if-conditions, so this would have to be `if (ret != 0)`. @nashif can confirm.

finikorg replied:
Ok, I see @nashif reaction, will change to boolean.

finikorg committed:
edac: shell: Replace ret with err

Replace ret with err for simple int type errors.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: shell: Use boolean expressions in if conditions

Use only boolean in if conditions as new coding style implies.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: ibecc: Use boolean expressions in if conditions

Use only boolean in if conditions as new coding style implies.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: ibecc: Simplify NMI handling

Refactor NMI handling making it clearer and fix return code for other
NMI source.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jhedberg commented on drivers/edac/edac_ibecc.c:
```diff
@@ -371,7 +386,7 @@ int edac_ibecc_init(const struct device *dev)
 	}
 
-	if (!ibecc_enabled(bdf)) {
+	if (ibecc_enabled(bdf) == false) {
```
If `ibecc_enabled()` returns `bool` then I don't think you need to do this modification (the bool type is a valid boolean statement in itself)

jhedberg commented on include/drivers/edac.h:
```diff
@@ -14,93 +14,77 @@
 		(const struct edac_driver_api *)dev->api;
 
-	return api->inject_set_param1(dev, addr);
+	if (!api->inject_set_param1) {
```
`if (api->inject_set_param == NULL)`?

jhedberg commented on include/drivers/edac.h:
```diff
@@ -109,48 +93,69 @@ static inline int edac_inject_set_param1(const struct device *dev,
 		(const struct edac_driver_api *)dev->api;
 
-	return api->inject_get_param1(dev);
+	if (!api->inject_get_param1) {
```
`if (api->inject_set_param == NULL)`?

jhedberg commented on include/drivers/edac.h:
```diff
@@ -109,48 +93,69 @@ static inline int edac_inject_set_param1(const struct device *dev,
 		(const struct edac_driver_api *)dev->api;
 
-	return api->inject_set_param2(dev, mask);
+	if (!api->inject_set_param2) {
```
`if (api->inject_set_param == NULL)`?

jhedberg commented on include/drivers/edac.h:
```diff
@@ -160,14 +165,20 @@ static inline uint64_t edac_inject_get_param2(const struct device *dev)
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
+	if (!api->inject_set_error_type) {
```
`if (api->inject_set_error_type == NULL)`?

jhedberg commented on include/drivers/edac.h:
```diff
@@ -177,122 +188,177 @@ static inline int edac_inject_set_error_type(const struct device *dev,
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
+	if (!api->inject_error_trigger) {
```
Same here

jhedberg commented on include/drivers/edac.h:
```diff
@@ -177,122 +188,177 @@ static inline int edac_inject_set_error_type(const struct device *dev,
 		(const struct edac_driver_api *)dev->api;
 
-	return api->ecc_error_log_get(dev);
+	if (!api->ecc_error_log_get) {
```
And here

jhedberg commented on include/drivers/edac.h:
```diff
@@ -177,122 +188,177 @@ static inline int edac_inject_set_error_type(const struct device *dev,
 		(const struct edac_driver_api *)dev->api;
 
-	api->ecc_error_log_clear(dev);
+	if (!api->ecc_error_log_clear) {
```
And here

jhedberg commented on include/drivers/edac.h:
```diff
@@ -177,122 +188,177 @@ static inline int edac_inject_set_error_type(const struct device *dev,
 		(const struct edac_driver_api *)dev->api;
 
-	return api->parity_error_log_get(dev);
+	if (!api->parity_error_log_get) {
```
And here

jhedberg commented on include/drivers/edac.h:
```diff
@@ -177,122 +188,177 @@ static inline int edac_inject_set_error_type(const struct device *dev,
 		(const struct edac_driver_api *)dev->api;
 
-	api->parity_error_log_clear(dev);
+	if (!api->parity_error_log_clear) {
```
And here

jhedberg commented on include/drivers/edac.h:
```diff
@@ -177,122 +188,177 @@ static inline int edac_inject_set_error_type(const struct device *dev,
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
+	if (!api->errors_cor_get) {
```
And here

jhedberg commented on include/drivers/edac.h:
```diff
@@ -177,122 +188,177 @@ static inline int edac_inject_set_error_type(const struct device *dev,
 	const struct edac_driver_api *api =
 		(const struct edac_driver_api *)dev->api;
 
+	if (!api->errors_uc_get) {
```
And here

jhedberg commented on include/drivers/edac.h:
```diff
@@ -303,13 +369,19 @@ static inline unsigned int edac_errors_uc_get(const struct device *dev)
 {
 	const struct edac_driver_api *api = dev->api;
 
+	if (!api->notify_cb_set) {
```
And here

finikorg committed:
edac: ibecc: Simplify PCI probing

Use only one pcie_conf_read() instead of several reads and make code
more readable.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: Use boolean expressions in if conditions

Use only boolean in if conditions as new coding style implies for edac
API header file.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jhedberg commented on include/drivers/edac.h:
```diff
@@ -109,48 +93,69 @@ static inline int edac_inject_set_param1(const struct device *dev,
 
-	return api->inject_get_param1(dev);
+	if (api->inject_get_param1 == NULL) {
+		return -ENOTSUP;
```
Sorry, I should have noticed this detail earlier. Whenever an API isn't implemented the right error is `ENOSYS` and not `ENOTSUP`. See here: https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/reference/api/design_guidelines.rst#return-codes

finikorg replied:
@jhedberg is `ENOSYS` is for mandatory API? This is optional API.

jhedberg replied:
@finikorg I don't think it makes any difference. Grep for ENOSYS under include/drivers to see how it's used.

finikorg committed:
edac: Use -ENOSYS also for optional interfaces

Use -ENOSYS for all not implemented interfaces.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: Add dumy EDAC driver for API coverage

Add dummy EDAC driver for -ENOSYS test coverage.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: ibecc: Strip debug information

Remove debug information increasing code coverage.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jhedberg commented on include/drivers/edac.h:
```diff
@@ -14,93 +14,80 @@
+ *
+ * @retval -ENOSYS if the optional interface is not implemented
+ * @retval 0 on success, other error code otherwise
  */
-static inline int edac_inject_set_param1(const struct device *dev,
-					 uint64_t addr)
+static inline int32_t edac_inject_set_param1(const struct device *dev,
```
The convention for APIs that return negative POSIX error numbers is `int` and not `int32`.

nashif replied:
what convention exactly? this convention you are talking about conflicts with the coding guidelines adopted by Zephyr.

jhedberg replied:
I mean't a de-facto convention, i.e. an established practice that can be seen throughout the tree. Whenever you see an `int` return you can assume that it has these semantics. I just counted 66 such APIs in `include/kernel.h` alone. If we move away from `int` I don't think `int32_t` is the right choice, since it looses essential semantic information  IMO. Instead we could introduce some new `err_t` type which is known to have these semantics.

jhedberg replied:
@finikorg I think you're referencing the wrong commit. In general, I think this has project-wide implications so review comments in this PR aren't necessary the right place for the discussion. I'd consider the API or Dev meetings instead.

finikorg replied:
@jhedberg Fixed commit.

carlescufi replied:
This has been raised before. I am of the opinion that using `int` wherever the size in bits of the variable doesn't matter is the correct approach, because it uses the native integer type of the platform. However, if this gets in the way of safety certification I agree we should consider it, and slowly transition to `int32_t` everywhere. In any case the error codes should always be documented with Doxygen so that it is clear for the user what to expect as a return value.
So to summarize my position: I'd rather use `int`, but safety certification is more important than my personal taste. If safety requires this then we need to transition to using `int32_t` everywhere for consistency, I assume then also on 64-bit platforms.

carlescufi replied:
@nashif wouldn't `size_t` be an issue as well in this case? since its size is platform-dependant, as @jhedberg observed offline.

jhedberg replied:
Btw, this also relates to #32213 , i.e. if we are going to change the commonly used error type we might want to introduce the concept of domain in the same go.

sjg20 replied:
I was looking here. Is it true that the rule has become mandatory in the 8 years from 2004 to 2012?

https://bullwhip.physio-control.com/documentation/help/reference/misra.builtin_numeric.htm

I agree with carlescufi on this point. 

Also, what do we do on 64-bit platforms with these APIs? Disallowing native types for error returns seems like a strange rule to me. You cannot stop an int32_t being promoted to a 64-bit value if that is the machine's register size. It might just add more confusion.

The page referenced above refers to variables that store some value with a meaning (speed, torque). I am not even sure that error codes fit that definition. If they do end up being stored in a data structure then using a specific length perhaps makes more sense.

finikorg replied:
I removed that patch introducing that change `int` -> `int32_t`. I did not like it at all, just had no other option ;)

finikorg committed:
edac: ibecc: Reduce CCM number using fallthrough

Reduce CCM number reported by Coverity by using __fallthrough.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
Rebased against latest upstream

jhedberg commented on drivers/edac/shell.c:
```diff
@@ -140,7 +164,13 @@ static int cmd_inject_mask(const struct shell *shell, size_t argc, char **argv)
+
+		err = edac_inject_get_param2(dev, &mask);
+		if (err != 0) {
+			shell_error(shell, "Error getting mask");
```
Maybe print the exact error as well with %d?

jhedberg replied:
I actually noticed that it's missing from many existing error cases, so a general improvement you could do as a separate patch

jhedberg commented on drivers/edac/shell.c:
```diff
@@ -150,10 +180,14 @@ static int cmd_inject_mask(const struct shell *shell, size_t argc, char **argv)
-		return edac_inject_set_param2(dev, value);
+		err = edac_inject_set_param2(dev, value);
+		if (err != 0) {
+			shell_error(shell, "Error setting mask");
```
Same here

finikorg committed:
edac: shell: Check error for log clear functions

Check errors returned by edac_ecc_error_log_clear() and
edac_parity_error_log_clear().

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: shell: Print exact error with shell error message

Include to shell_error() also exact error.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: shell: Use helper functions for shell devmem command

Using helper functions decreases function complexity numbers reported
by static code analyzers.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jhedberg commented on drivers/edac/shell.c:
```diff
@@ -62,39 +62,51 @@ static int cmd_edac_info(const struct shell *shell, size_t argc, char **argv)
-	error = edac_parity_error_log_get(dev);
+	err = edac_parity_error_log_get(dev, &error);
+	if (err != 0) {
+		shell_error(shell, "Error getting parity error log");
```
Missing err %d

finikorg replied:
Sorry, missed these, fixed now

jhedberg commented on drivers/edac/shell.c:
```diff
@@ -62,39 +62,51 @@ static int cmd_edac_info(const struct shell *shell, size_t argc, char **argv)
-	error = edac_ecc_error_log_get(dev);
+	err = edac_ecc_error_log_get(dev, &error);
+	if (err != 0) {
+		shell_error(shell, "Error getting ecc error log");
```
Missing err %d

jhedberg reviewed (approved):
Looks fine, except for a couple of places with missing `err %d`
